src/tools/r2rdump/R2RDump.cs

```diff
@@ -4,55 +4,432 @@
+        public IList<R2RMethod> FindMethod(R2RReader r2r, string query, bool exact)
+        {
+            List<R2RMethod> res = new List<R2RMethod>();
+            foreach (R2RMethod method in r2r.R2RMethods)
+            {
+                if (Match(method, query, exact))
+                {
+                    res.Add(method);
+                }
+            }
+            return res;
+        }
+
+        /// <summary>
+        /// Finds all R2RSections by name or value of the ReadyToRunSectionType matching <param>query</param>
+        /// </summary>
+        /// <param name="r2r">Contains all extracted info about the ReadyToRun image</param>
+        /// <param name="query">The name or value to search for</param>
+        /// <out name="res">List of all matching sections</out>
+        /// <remarks>Case-insensitive</remarks>
+        public IList<R2RSection> FindSection(R2RReader r2r, string query)
+        {
+            List<R2RSection> res = new List<R2RSection>();
+            foreach (R2RSection section in r2r.R2RHeader.Sections.Values)
+            {
+                if (Match(section, query))
+                {
+                    res.Add(section);
+                }
+            }
+            return res;
+        }
+
+        /// <summary>
+        /// Returns the runtime function with id matching <param>rtfQuery</param>
+        /// </summary>
+        /// <param name="r2r">Contains all extracted info about the ReadyToRun image</param>
+        /// <param name="rtfQuery">The name or value to search for</param>
+        public RuntimeFunction FindRuntimeFunction(R2RReader r2r, int rtfQuery)
+        {
+            foreach (R2RMethod m in r2r.R2RMethods)
+            {
+                foreach (RuntimeFunction rtf in m.RuntimeFunctions)
+                {
+                    if (rtf.Id == rtfQuery || (rtf.StartAddress >= rtfQuery && rtf.StartAddress + rtf.Size < rtfQuery))
                     {
-                        Console.Write(method.ToString());
-                        Console.WriteLine("------------------\n");
+                        return rtf;
                     }
                 }
             }
+            return null;
+        }
+
+        private int Run(string[] args)
+        {
+            ArgumentSyntax syntax = ParseCommandLine(args);
+            
+            if (_help)
+            {
+                Console.WriteLine(syntax.GetHelpText());
+                return 0;
+            }
+
+            if (_inputFilenames.Count == 0)
+                throw new ArgumentException("Input filename must be specified (--in <file>)");
+
+            // open output file
+            FileStream fileStream = null;
+            StreamWriter writer = null;
+            if (_outputFilename != null)
+            {
+                fileStream = new FileStream(_outputFilename, FileMode.Create, FileAccess.Write);
+                writer = new StreamWriter(fileStream);
+                Console.SetOut(writer);
```

Comment:
I think something that would be nice to have is to not use the Console.SetOut method, but instead have a new sort of Writer object that takes a TextStream in its constructor and always writes to that. That way we can easily change what/where it is writing to by changing the argument passed into the constructor. That's not something that needs to happen in this PR though.